src/src/alarm.c

```diff
@@ -50,31 +50,90 @@
 #include "rail.h"
 #include "sl_sleeptimer.h"
 
+// timer data for handling wrapping
+typedef struct wrap_timer_data wrap_timer_data_t;
+struct wrap_timer_data
+{
+    uint16_t overflow_counter;
+    uint16_t overflow_max;
+};
+
 // millisecond timer (sleeptimer)
 static sl_sleeptimer_timer_handle_t sl_handle;
-static uint32_t                     sMsAlarm     = 0;
-static bool                         sIsMsRunning = false;
+static uint64_t                     sMsAlarm         = 0;
+static bool                         sIsMsRunning     = false;
+static wrap_timer_data_t            milli_timer_data = {0};
 
 // microsecond timer (RAIL timer)
-static uint32_t sUsAlarm     = 0;
-static bool     sIsUsRunning = false;
-
 static RAIL_MultiTimer_t rail_timer;
+static uint64_t          sUsAlarm         = 0;
+static bool              sIsUsRunning     = false;
+static wrap_timer_data_t micro_timer_data = {0};
 
+// millisecond-alarm callback
 static void AlarmCallback(sl_sleeptimer_timer_handle_t *aHandle, void *aData)
 {
-    OT_UNUSED_VARIABLE(aHandle);
-    OT_UNUSED_VARIABLE(aData);
-    otSysEventSignalPending();
+    if (aData == NULL)
+    {
+        OT_UNUSED_VARIABLE(aHandle);
+        otSysEventSignalPending();
+    }
+    else
+    {
+        sl_status_t        status;
+        wrap_timer_data_t *timer_data = (wrap_timer_data_t *)aData;
+
+        if (timer_data->overflow_counter < timer_data->overflow_max)
+        {
+            status = sl_sleeptimer_start_timer_ms(aHandle,
+                                                  sl_sleeptimer_get_max_ms32_conversion(),
+                                                  AlarmCallback,
+                                                  (void *)timer_data,
+                                                  0,
+                                                  SL_SLEEPTIMER_NO_HIGH_PRECISION_HF_CLOCKS_REQUIRED_FLAG);
+            OT_ASSERT(status == SL_STATUS_OK);
+            timer_data->overflow_counter++;
+        }
+        else
+        {
+            if (timer_data->overflow_max != 0)
+            {
+                sIsMsRunning = false;
+                sl_sleeptimer_stop_timer(aHandle);
+            }
+        }
+    }
 }
 
+// microsecond-alarm callback
 static void radioTimerExpired(struct RAIL_MultiTimer *tmr, RAIL_Time_t expectedTimeOfEvent, void *cbArg)
 {
-    OT_UNUSED_VARIABLE(tmr);
-    OT_UNUSED_VARIABLE(expectedTimeOfEvent);
-    OT_UNUSED_VARIABLE(cbArg);
+    if (cbArg == NULL)
+    {
+        OT_UNUSED_VARIABLE(tmr);
+        OT_UNUSED_VARIABLE(expectedTimeOfEvent);
+        otSysEventSignalPending();
+    }
+    else
+    {
+        RAIL_Status_t      status;
+        wrap_timer_data_t *timer_data = (wrap_timer_data_t *)cbArg;
 
-    otSysEventSignalPending();
+        if (timer_data->overflow_counter < timer_data->overflow_max)
+        {
```

Comment:
Not for this PR, but I think we can combine the two timer expired callbacks.

Reply:
I agree, but I also have that as an action item for myself with planned alarm updates for renode support, so please don't duplicate any work!